viamonvif/device/device.go

```diff
@@ -0,0 +1,326 @@
+	for i, profile := range resp.Body.GetProfilesResponse.Profiles {
+		dev.logger.Debugf("Profile %d: Token=%s, Name=%s", i, profile.Token, profile.Name)
+	}
+
+	dev.logger.Debugf("GetProfiles decoded: %#v", resp.Body.GetProfilesResponse)
+	return resp.Body.GetProfilesResponse, nil
+}
+
+type getStreamURIResponse struct {
+	XMLName xml.Name `xml:"Envelope"`
+	Body    struct {
+		GetStreamURIResponse struct {
+			MediaURI onvif.MediaUri `xml:"MediaUri"`
+		} `xml:"GetStreamUriResponse"`
+	} `xml:"Body"`
+}
+
+// Credentials contain an onvif device username and password.
+type Credentials struct {
+	User string `json:"user"`
+	Pass string `json:"pass"`
+}
+
+// GetStreamURI returns a device's stream URI for a given profile.
+func (dev *Device) GetStreamURI(profile onvif.Profile, creds Credentials) (*url.URL, error) {
+	dev.logger.Debugf("GetStreamUri token: %s, profile: %#v", profile.Token, profile)
+	body, err := dev.callMedia(GetStreamURI{
+		StreamSetup: onvif.StreamSetup{
+			Stream:    onvif.StreamType(streamTypeRTPUnicast),
+			Transport: onvif.Transport{Protocol: streamSetupProtocol},
+		},
+		ProfileToken: profile.Token,
+	})
+	if err != nil {
+		return nil, err
+	}
+	dev.logger.Debugf("GetStreamUri response: %v", string(body))
+
+	var streamURI getStreamURIResponse
+	if err := xml.NewDecoder(bytes.NewReader(body)).Decode(&streamURI); err != nil {
+		return nil, fmt.Errorf("failed to get RTSP URL for profile %s: %w", profile.Token, err)
+	}
+
+	dev.logger.Debugf("GetStreamUriResponse decoded %v: ", profile.Token, streamURI)
+
+	uriStr := string(streamURI.Body.GetStreamURIResponse.MediaURI.Uri)
+	if uriStr == "" {
+		return nil, fmt.Errorf("got empty uri for profile %s", profile.Token)
+	}
+
+	uri, err := url.Parse(uriStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI %s: %w", uriStr, err)
+	}
+
+	if creds.User != "" || creds.Pass != "" {
+		uri.User = url.UserPassword(creds.User, creds.Pass)
+	}
+	return uri, nil
+}
+
+// GetEndpoint returns specific ONVIF service endpoint address.
+func (dev *Device) GetEndpoint(name string) string {
+	return dev.endpoints[name]
+}
+
+func (dev Device) callMedia(method interface{}) ([]byte, error) {
+	return dev.callMethodDo(dev.endpoints["media"], method)
+}
+
+func (dev Device) callDevice(method interface{}) ([]byte, error) {
+	return dev.callMethodDo(dev.endpoints["device"], method)
+}
+
+func (dev Device) callMethodDo(endpoint string, method interface{}) ([]byte, error) {
+	output, err := xml.MarshalIndent(method, "  ", "    ")
+	if err != nil {
+		return nil, err
+	}
+
+	doc := etree.NewDocument()
+	if err := doc.ReadFromBytes(output); err != nil {
+		return nil, err
+	}
+
+	soap, err := gosoap.NewEmptySOAP()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := soap.AddBodyContent(doc.Root()); err != nil {
+		return nil, err
+	}
+	for key, value := range Xlmns {
+		if err := soap.AddRootNamespace(key, value); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Is there any instance where we would want to ignore one of the erroring xlmns and continue to try the others?

Reply:
in this case no. 

the XML being encoded in this path always comes from our code, not from the device.

As a consequence this code would only error if we make a mistake at compile time.

This is one of the reasons I would remove the dependence on the `etree` library as it has an API that is optimized for untrusted xmls, which adds the need to handle a bunch of errors which are impossible in our application b/c we are creating the xmls by hand at compile time.